clients/drcachesim/tests/raw2trace_io.cpp

```diff
@@ -0,0 +1,161 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This application processes a set of existing raw files with raw2trace_t.
+ * It uses ptrace to make sure raw2trace only interacts with the filesystem when expected
+ * -- it doesn't open/close any files outside of mapping modules (which should be local).
+ *
+ * XXX: We use ptrace as opposed to running under memtrace with replaced file operations
+ * because raw2trace uses drmodtrack, which doesn't isolate under static memtrace.
+ */
+
+#include "droption.h"
+#include "tracer/raw2trace.h"
+#include "tracer/raw2trace_directory.h"
+#include <cassert>
+#include <errno.h>
+#include <fcntl.h>
+#include <iostream>
+#include <string>
+#include <sys/ptrace.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+#include <syscall.h>
+#include <unistd.h>
+
+static droption_t<std::string> op_indir(DROPTION_SCOPE_FRONTEND, "indir", "",
+                                        "[Required] Directory with trace input files",
+                                        "Specifies a directory with raw files.");
+
+static droption_t<std::string> op_out(DROPTION_SCOPE_FRONTEND, "out", "",
+                                      "[Required] Path to output file",
+                                      "Specifies the path to the output file.");
+#if defined(X64)
+# define SYS_NUM(r) (r).orig_rax
+#elif defined(X86)
+# define SYS_NUM(r) (r).orig_eax
+#else
+# error "Test only supports x86"
+#endif
+
+int
+main(int argc, const char *argv[])
+{
+    std::string parse_err;
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, (const char **)argv,
+                                       &parse_err, NULL) ||
+        op_indir.get_value().empty() || op_out.get_value().empty()) {
+        std::cerr << "Usage error: " << parse_err << "\nUsage:\n"
+                  << droption_parser_t::usage_short(DROPTION_SCOPE_ALL);
+        return 1;
+    }
+
+    /* Open input/output files outside of traced region. And explicitly don't destroy dir,
+     * so they never get closed.
+     */
+    raw2trace_directory_t *dir =
+        new raw2trace_directory_t(op_indir.get_value(), op_out.get_value());
```

Comment:
Hmm, ok, it's not quite what I was thinking of (which was more of a test that actually uses the new i/o interface), but it does test the i/o isolation.